harness/determined/keras/_callback.py

```diff
@@ -0,0 +1,446 @@
+
+        self._is_chief = core_context.distributed.rank == 0
+
+        # This is an undocumented workaround in case off-cluster user-code saving goes awry.
+        self._save_user_code = True
+
+        self._steps_completed = 0
+        self._last_ckpt_epoch = -1
+        self._last_train_epoch = -1
+
+        # progress
+        self._training_length: Optional[int] = None
+        self._validation_length: Optional[int] = None
+        self._training_batches = 0
+        self._validation_batches = 0
+        self._percent_reported = -1
+
+        # We download the checkpoint, then have to keep it for a while until we can delete it.
+        self._ckpt_context: Optional[contextlib.ExitStack] = None
+
+    # Mask the inherited ProgbarLogger behavior that we don't actually want.
+    def set_params(self, params: Dict[str, Any]) -> None:
+        callbacks.Callback.set_params(self, params)
+
+    # Mask the inherited ProgbarLogger behavior that we don't actually want.
+    def on_predict_begin(self, logs: Optional[Dict[str, Any]]) -> None:
+        pass
+
+    # Mask the inherited ProgbarLogger behavior that we don't actually want.
+    def on_predict_batch_end(self, batch: int, logs: Optional[Dict[str, Any]]) -> None:
+        pass
+
+    # Mask the inherited ProgbarLogger behavior that we don't actually want.
+    def on_predict_end(self, logs: Optional[Dict[str, Any]]) -> None:
+        pass
+
+    def _implements_train_batch_hooks(self) -> bool:
+        return True
+
+    def _implements_test_batch_hooks(self) -> bool:
+        # Tell keras that we don't need on_test_batch_end unless we are verbose.
+        return bool(self.params["verbose"] != 0)
+
+    def _implements_predict_batch_hooks(self) -> bool:
+        # Tell keras that we don't actually want any on_predict_batch_end calls.
+        return False
+
+    def _print_progress(
+        self, logs: Optional[Dict[str, Any]], training: bool, batches: int, total: Optional[int]
+    ) -> None:
+        # Only report progress if we have the target total.
+        if total is None:
+            return
+
+        # Don't report more often than 10% increments.
+        percent_10 = int((batches / total) * 10) * 10
+        if percent_10 <= self._percent_reported:
+            return
+
+        # When you do report, report to 1% accuracy.
+        percent = int((batches / total) * 100)
+        self._percent_reported = percent
+
+        if training:
+            report = (
+                f"total batches trained: {self._steps_completed}, "
+                f"epoch {percent}% complete ({batches}/{total})"
+            )
+        else:
+            report = (
+                f"validation after batch: {self._steps_completed}, "
+                f"validation {percent}% complete ({batches}/{total})"
+            )
+        if logs is not None:
+            metrics = {k: v for k, v in logs.items() if k not in ("batch", "size")}
+            report += f": {metrics}"
```

Comment:
nit: newline for metrics `f":\n {metrics}"`?

Reply:
I think having it on one line is much nicer to look at.